pkg/castemplate/v1alpha1/kubernetes_test.go

```diff
@@ -0,0 +1,213 @@
+		// tested using these two combinations only.
+		"When mockclient is empty": {nil, nil, false, false},
+		"When mockclient contains all of them": {fakeGetfn,
+			fakeGetClientset, false, false},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			fc := &Kubeclient{}
+			fc.get = mock.getFn
+			fc.getClientset = mock.getClientsetFn
+
+			fc.withDefaults()
+			get := (fc.get == nil)
+			if get != mock.expectGet {
+				t.Fatalf(`test %s failed: expected non-nil fc.get
+but got %v`, name, fc.get)
+			}
+			getClientset := (fc.getClientset == nil)
+			if getClientset != mock.expectGetClientset {
+				t.Fatalf(`test %s failed: expected non-nil fc.getClientset
+but got %v`, name, fc.getClientset)
+			}
+		})
+	}
+}
+func TestWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		clientSet    *clientset.Clientset
+		isKubeClient bool
+	}{
+		"Clientset is empty":     {nil, false},
+		"Clientset is not empty": {&clientset.Clientset{}, true},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			h := WithClientset(mock.clientSet)
+			fake := &Kubeclient{}
+			h(fake)
+			if mock.isKubeClient && fake.clientset == nil {
+				t.Fatalf(`test %s failed, expected non-nil fake.clientset
+but got %v`, name, fake.clientset)
+			}
+			if !mock.isKubeClient && fake.clientset != nil {
+				t.Fatalf(`test %s failed, expected nil fake.clientset
+but got %v`, name, fake.clientset)
+			}
+		})
+	}
+}
+func TestKubeClientWithClientset(t *testing.T) {
+	tests := map[string]struct {
+		expectClientSet bool
+		opts            []KubeclientBuildOption
+	}{
+		"When non-nil clientset is passed": {true,
+			[]KubeclientBuildOption{fakeSetClientset}},
+		"When two options with a non-nil clientset are passed": {true,
+			[]KubeclientBuildOption{fakeSetClientset, fakeClientSet}},
+		"When three options with a non-nil clientset are passed": {true,
+			[]KubeclientBuildOption{fakeSetClientset, fakeClientSet, fakeClientSet}},
+
+		"When nil clientset is passed": {false,
+			[]KubeclientBuildOption{fakeSetNilClientset}},
+		"When two options with a nil clientset are passed": {false,
+			[]KubeclientBuildOption{fakeSetNilClientset, fakeClientSet}},
+		"When three options with a nil clientset are passed": {false,
+			[]KubeclientBuildOption{fakeSetNilClientset, fakeClientSet, fakeClientSet}},
+	}
+
+	for name, mock := range tests {
+		t.Run(name, func(t *testing.T) {
+			c := KubeClient(mock.opts...)
+			if !mock.expectClientSet && c.clientset != nil {
+				t.Fatalf(`test %s failed, expected nil c.clientset
+but got %v`, name, c.clientset)
+			}
+			if mock.expectClientSet && c.clientset == nil {
+				t.Fatalf(`test %s failed expected non-nil c.clientset
+but got %v`, name, c.clientset)
+			}
+		})
+	}
+}
+
+func TestGetClientOrCached(t *testing.T) {
+	tests := map[string]struct {
+		kubeClient *Kubeclient
+		expectErr  bool
+	}{
+		// Positive tests
+		"When clientset is nil": {&Kubeclient{nil,
```

Comment:
Each field of Kubeclient that is used here should be passed as `inputs` in above struct

Reply:
done